src/gui/optionsdialog.cpp

```diff
@@ -1648,57 +1645,76 @@ int OptionsDialog::getActionOnDblClOnTorrentFn() const
-            break;
-        case ScanFoldersModel::DoesNotExist:
-            error = tr("Folder does not exist:");
-            break;
-        case ScanFoldersModel::CannotRead:
-            error = tr("Folder is not readable:");
-            break;
-        default:
-            pref->setScanDirsLastPath(dir);
-            m_addedScanDirs << dir;
-            for (int i = 0; i < ScanFoldersModel::instance()->columnCount(); ++i)
-                m_ui->scanFoldersView->resizeColumnToContents(i);
-            enableApplyButton();
-        }
+        auto watchedFoldersModel = static_cast<WatchedFoldersModel *>(m_ui->scanFoldersView->model());
+        watchedFoldersModel->addFolder(dir, dialog.watchedFolderOptions());
 
-        if (!error.isEmpty())
-            QMessageBox::critical(this, tr("Adding entry failed"), QString::fromLatin1("%1\n%2").arg(error, dir));
+        pref->setScanDirsLastPath(dir);
+
+        for (int i = 0; i < watchedFoldersModel->columnCount(); ++i)
+            m_ui->scanFoldersView->resizeColumnToContents(i);
+
+        enableApplyButton();
+    }
+    catch (const RuntimeError &err)
+    {
+        QMessageBox::critical(this, tr("Adding entry failed"), err.message());
     }
 }
 
-void OptionsDialog::on_removeScanFolderButton_clicked()
+void OptionsDialog::on_editWatchedFolderButton_clicked()
+{
+    const QModelIndex selected
+        = m_ui->scanFoldersView->selectionModel()->selectedIndexes().at(0);
+
+    editWatchedFolderOptions(selected);
+}
+
+void OptionsDialog::on_removeWatchedFolderButton_clicked()
 {
     const QModelIndexList selected
         = m_ui->scanFoldersView->selectionModel()->selectedIndexes();
-    if (selected.isEmpty())
-        return;
-    Q_ASSERT(selected.count() == ScanFoldersModel::instance()->columnCount());
+
     for (const QModelIndex &index : selected)
-    {
-        if (index.column() == ScanFoldersModel::WATCH)
-            m_removedScanDirs << index.data().toString();
-    }
-    ScanFoldersModel::instance()->removePath(selected.first().row(), false);
+        m_ui->scanFoldersView->model()->removeRow(index.row());
+}
+
+void OptionsDialog::handleWatchedFolderViewSelectionChanged()
+{
+    m_ui->removeWatchedFolderButton->setEnabled(!m_ui->scanFoldersView->selectionModel()->selectedIndexes().isEmpty());
+    m_ui->editWatchedFolderButton->setEnabled(m_ui->scanFoldersView->selectionModel()->selectedIndexes().count() == 1);
 }
 
-void OptionsDialog::handleScanFolderViewSelectionChanged()
+void OptionsDialog::editWatchedFolderOptions(const QModelIndex &index)
 {
-    m_ui->removeScanFolderButton->setEnabled(!m_ui->scanFoldersView->selectionModel()->selectedIndexes().isEmpty());
+    if (!index.isValid())
+        return;
+
+    auto watchedFoldersModel = static_cast<WatchedFoldersModel *>(m_ui->scanFoldersView->model());
+    WatchedFolderOptionsDialog dialog {watchedFoldersModel->folderOptions(index.row()), this};
+    if (dialog.exec() == QDialog::Accepted)
```

Comment:
The dialog should be invoked in async way to avoid other problems.

Reply:
>The dialog should be invoked in async way to avoid other problems.

Done.

Reply:
Oops... I fixed it in one place, but forgot to do it in another.